File: haskell/glue/src/Glue/Lib/Math/Trigonometric/Acos.hs
module Glue.Lib.Math.Trigonometric.Acos where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Arccosine function
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Acos.acos exactly
acos :: IR Eval
acos = NativeFunc acosImpl

-- Arccosine function implementation
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Acos.acosImpl exactly
acosImpl :: IR Eval -> Eval (IR Eval)
acosImpl arg = case arg of
    Integer n -> pure $ Float (Prelude.acos (fromIntegral n))
    Float n -> pure $ Float (Prelude.acos n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/trigonometric/acos.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Arccosine function (returns radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Acos.acos exactly
final Ir acos = IrNativeFunc(acosImpl);

/// Arccosine function implementation (returns radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Acos.acosImpl exactly
Eval<Ir> acosImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(IrFloat(math.acos(n.toDouble()))),
    IrFloat(value: final n) => Eval.pure(IrFloat(math.acos(n))),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
